pkg/spoke/controllers/finalizercontroller/manifestwork_finalize_controller.go

```diff
@@ -0,0 +1,139 @@
+	worklister "github.com/open-cluster-management/api/client/work/listers/work/v1"
+	"github.com/open-cluster-management/work/pkg/helper"
+	"github.com/open-cluster-management/work/pkg/spoke/controllers"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+)
+
+// ManifestWorkFinalizeController handles cleanup of manifestwork resources before deletion is allowed.
+type ManifestWorkFinalizeController struct {
+	manifestWorkClient        workv1client.ManifestWorkInterface
+	manifestWorkLister        worklister.ManifestWorkNamespaceLister
+	appliedManifestWorkClient workv1client.AppliedManifestWorkInterface
+	appliedManifestWorkLister worklister.AppliedManifestWorkLister
+	hubHash                   string
+	rateLimiter               workqueue.RateLimiter
+}
+
+func NewManifestWorkFinalizeController(
+	recorder events.Recorder,
+	manifestWorkClient workv1client.ManifestWorkInterface,
+	manifestWorkInformer workinformer.ManifestWorkInformer,
+	manifestWorkLister worklister.ManifestWorkNamespaceLister,
+	appliedManifestWorkClient workv1client.AppliedManifestWorkInterface,
+	appliedManifestWorkInformer workinformer.AppliedManifestWorkInformer,
+	hubHash string,
+) factory.Controller {
+
+	controller := &ManifestWorkFinalizeController{
+		manifestWorkClient:        manifestWorkClient,
+		manifestWorkLister:        manifestWorkLister,
+		appliedManifestWorkClient: appliedManifestWorkClient,
+		appliedManifestWorkLister: appliedManifestWorkInformer.Lister(),
+		hubHash:                   hubHash,
+		rateLimiter:               workqueue.NewItemExponentialFailureRateLimiter(5*time.Millisecond, 1000*time.Second),
+	}
+
+	return factory.New().
+		WithInformersQueueKeyFunc(func(obj runtime.Object) string {
+			accessor, _ := meta.Accessor(obj)
+			return accessor.GetName()
+		}, manifestWorkInformer.Informer()).
+		WithInformersQueueKeyFunc(helper.AppliedManifestworkQueueKeyFunc(hubHash), appliedManifestWorkInformer.Informer()).
+		WithSync(controller.sync).ToController("ManifestWorkFinalizer", recorder)
+}
+
+func (m *ManifestWorkFinalizeController) sync(ctx context.Context, controllerContext factory.SyncContext) error {
+	manifestWorkName := controllerContext.QueueKey()
+	klog.V(4).Infof("Reconciling ManifestWork %q", manifestWorkName)
+
+	manifestWork, err := m.manifestWorkLister.Get(manifestWorkName)
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+
+	// If manifestwork is found, check if it is being deleted and has the finalizer.
+	if err == nil {
+		// no work to do until we're deleted
+		if manifestWork.DeletionTimestamp == nil || manifestWork.DeletionTimestamp.IsZero() {
+			return nil
+		}
+
+		// don't do work if the finalizer is not present
```

Comment:
I think that you need to delete the corresponding AppliedManifestWork if the ManifestWork is deleted even if the finalizer is gone.

Reply:
done